### PR TITLE
Release

### PR DESCRIPTION
**Pull request by kensayo**
Project 2: Enumerable Methods

We created the basic enumerable methods from scratch to work as defined Enumerable Methos provided by Ruby

- Create #my_each method.
- Create #my_each_with_index method.
- Create #my_select  method.
- Create #my_all?  method.
- Create #my_any? method.
- Create #my_none? method.
- Create #my_count method.
- Create #my_map method.
- Create #my_inject method.
- Create #multiply_els to receive an array as an argument.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -1 +1,173 @@
...
+      elsif args
+        to_a.my_each { |var| return false unless my_instance_of(args, var) }
+      end
+      to_a.my_each { |var| return false unless var }
+      return true
+    end
+    to_a.size.times { |i| return false unless yield to_a[i] }
+
+    true
+  end
```
Comment (george-shammar) [inline main.rb:71]: - [ ]  When a pattern other than Regex or a Class is given, it should return true if all of the collection matches the pattern.

File: main.rb [hunk 2]
```diff
@@ -1 +1,173 @@
...
+      elsif args
+        to_a.my_each { |var| return true if my_instance_of(args, var) }
+      end
+
+      to_a.my_each { |var| return true if var == true }
+      return false
+    end
+    to_a.my_each { |var| return true if yield var }
+    false
+  end
```
Comment (george-shammar) [inline main.rb:88]: - [ ] When a pattern other than Regex or a Class is given, it should return true if any of the collection matches the pattern.

File: main.rb [hunk 3]
```diff
@@ -1 +1,173 @@
...
+        return true
+      elsif args
+        to_a.my_each { |var| return false if my_instance_of(args, var) }
+      end
+      to_a.my_each { |var| return false if var }
+      return true
+    end
+    my_each { |var| return false if yield var }
+    true
+  end
```
Comment (george-shammar) [inline main.rb:104]: - [ ] When a pattern other than Regex or a Class is given, it should return true only if none of the collection matches the pattern.

File: main.rb [hunk 4]
```diff
@@ -1 +1,173 @@
...
+        counter += 1 if var == num
+      end
+
+    else
+      my_each do |var|
+        counter += 1 if yield var
+      end
+    end
+    counter
+  end
```
Comment (george-shammar) [inline main.rb:122]: - [ ] It should count the number of items in enum that are equal to item if an argument is given.

File: main.rb [hunk 5]
```diff
@@ -1 +1,173 @@
...
+    my_each { |var| mp_new.push(yield(var)) }
+    mp_new
+  end
+
+  def my_instance_of(args, var)
+    return true if var.class == args || var.class.superclass == args
+
+    false
+  end
+end
```
Comment (george-shammar) [inline main.rb:138]: - [ ] It should execute only the proc when both a block and a proc are given and it should return a new array with the results of running the proc once for every element in enum.

File: main.rb [hunk 6]
```diff
@@ -1 +1,173 @@
...
+p [nil, false].my_none? #=> true
+p [nil, false, true].my_none? #=> false
+p 'count examples' # Count
+ary = [1, 2, 4, 2]
+p ary.count(&:even?) #=> 3
+p ary.count #=>4
+p ary.count(2) #=>2
+p 'map examples' # MAP
+p [1, 2, 3, 4].my_map { |i| i * i } #=> [1, 4, 9, 16]
+p [1, 2, 3, 4].my_map { 'cat' } #=> ["cat", "cat", "cat", "cat"]
```
Comment (george-shammar) [inline main.rb:171]: - [ ] All these lines of code are unnecessary. Please remove them.

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUESTED** :recycle: :recycle: 

**Wow!!!** **This is a very strong and impressive first submission.**  :clap: :clap: :+1: 

Thanks for the hard work you have put into this. Good job, guys. :+1: 

Let's make small improvements to get the project approved. You need to update your methods to cover some edge cases yet to be covered.

Please see my inline comments for each enumerable method.

**Generally:**

- [x] Please include multiply_els method according to the project requirement, which should accept an array as an argument and multiply all the elements of the array together using #my_inject.
- [ ] In your pull request message, you have simply copied and pasted the project instruction. Please summarize the changes you're proposing to merge to the default branch in your pull request message.

Please submit another CRR when you're ready.

HAPPY CODING! :+1:

**Comment by mariordgez** [comment]
Here are the changes that we added for the next code review:

- Included multiply_els method, that works as project specification requires
- Made changes to the pull request as a whole
- The methods my_any, my_none and my_all now works with other kinds of patterns outside of regexp and class
- The my_count method now counts the number of items in enum that are equal to item if an argument is given
- The my_map method now correctly takes a proc, and if given both a proc and a block, it prioritizes the proc as the project requires

I hope you like the project and have a nice day!

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
# Status: Request changes

Your project looks wow guys @kensayo @mariordgez almost all your enumerable methods are working perfectly and you implemented almost all changes requested by the TSE but you need to make a small change and you are ready to move on

## To highlights image

- You nailed your PR it shows professionalism
- Git workflow is dope

## To improve image

### my_none

`%w[dog door rod blade].my_none?(5)`

- enumerable #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

## Kindly fix that and submit another CRR

Regards,

Lydia Ingabire

Slack; Delice Lydia
[Linkedin](https//www.linkedin.com/in/delice-lydia/)
[Twitter](https://twitter.com/IngabireLydia3)

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:

### Congratulations! Thank you for implementing changes suggested previously 🙏
### You can now merge your branch with the main branch!

## "Highly Recommended changes"

### Please do not push unnecessary files and folders to your repo. You can add them in gitignore if they are so important for you to work locally!

### Happy Coding!

## You can always connect with me on slack ( @Sajjad Ahmad ).

**Pull request merged**